pkgs/development/beam-modules/build-mix.nix

```diff
@@ -0,0 +1,89 @@
+{ stdenv, writeText, elixir, erlang, hex, lib }:
+
+{ name
+, version
+, src
+, buildInputs ? [ ]
+, beamDeps ? [ ]
+, propagatedBuildInputs ? [ ]
+, postPatch ? ""
+, compilePorts ? false
+, meta ? { }
+, enableDebugInfo ? false
+, mixEnv ? "prod"
+, ...
+}@attrs:
+
+with lib;
+let
+  shell = drv: stdenv.mkDerivation {
+    name = "interactive-shell-${drv.name}";
+    buildInputs = [ drv ];
+  };
+
+  pkg = self: stdenv.mkDerivation (attrs // {
+    name = "${name}-${version}";
+    inherit version;
+    inherit src;
+
+    MIX_ENV = mixEnv;
+    MIX_DEBUG = if enableDebugInfo then 1 else 0;
+    HEX_OFFLINE = 1;
+
+    setupHook = attrs.setupHook or
+      writeText "setupHook.sh" ''
```

Comment:
I would probably add a comment here like
```
# $ERL_LIB is how mix finds dependencies
[[ $MIX_DEBUG -eq 1 ]] && echo "Adding $1/lib/erlang/lib to $ERL_LIB"
```

Reply:
One more thought here. (the ERL_LIBS thing needs to be clarified, but leaving that aside)

instead of parsing the ERL_LIBS again in the configurePhase, how about populating the _build directory here directly ?
something like
```
      mkdir -p _build/$MIX_ENV/lib
      lib_path="$1/lib/erlang/lib"
      addToSearchPath ERL_LIBS "$lib_path"
      for dir in $(ls $lib_path) ; do
        dest=$(echo "$dir" | cut -d '-' -f1)
        # some dependencies could have populated the path
        # makeup_elixir depends on makeup and nimble_parsec, makeup also depends on nimble_parsec
        if ! test -d _build/$MIX_ENV/lib/$dest ; then
          [[ $MIX_DEBUG -eq 1 ]] && echo "Linking $lib_path/$dir to _build/$MIX_ENV/lib/$dest"
          ln -s $lib_path/$dir _build/$MIX_ENV/lib/$dest
        fi
      done
```

Just a thought, this adds an if condition but saves the `while read` part

Reply:
The setup hook executes when setting up the environment. It is not only executed when building. It is also executed whenever the environment is constructed ( for example, when using `nix-shell` ). The configurePhase is part of the build process. They should remain separate.

Reply:
Thanks for the clarification!